### PR TITLE
textures for 2.80

### PR DESCRIPTION
**Pull request by Grodou**
Simple image texture export
Image file copy (direct and packed)
Fixes hidden object option

Only tested on Linux.

Changes and question
----------------------
util.py
----------------------
should_export()
	Tells if an object should be exported (uses bpy.context.visible_objects)
	Should we use new visibility modes in 2.80:
		obj.hide_render
		obj.hide_select
		obj.hide_viewport
		
IndentedWriter
	inserts a space before '{'
	removes the extra space at the end of line
	should not break anything

----------------------
ogre/scene.py
----------------------
	Adds ".material" extension to the material file name
	Does not work for me otherwise

----------------------
ogre/material.py
----------------------
Should we keep change_ext ?
generate
	images are collected while generating passes, so the collect action is only made once, as well as the file copy
	is normal map useful ?
	Why was the rotation based on translation.z value ?
		> tried to use the rotation.z value (in radians) instead, but don't know how to use it properly
	Why is projection (SPHERE/FLAT) only used for 'Reflection' coordinates ?
	
copy_textures
	image copy now uses shutil.copy2 to keep metadata, update is made if at least sizes or modification dates are different
	Assuming images are always copied to the target folder with no hierarchy ('TOUCH_TEXTURES' option is not available in the UI), if 2 files have the same name, only one will remain in the export folder
		> use texture name or label instead ?

### CODE REVIEW

File: io_ogre/ogre/material.py [hunk 1]
```diff
@@ -175,11 +202,57 @@ def generate_pass( self, mat, pass_name="" ):
...
+        """
+        src_dir = os.path.dirname(bpy.data.filepath)
+        # For target path relative
+        # dst_dir = os.path.dirname(self.target_path)
+        dst_dir = src_dir
+        filename = io_utils.path_reference(texture.image.filepath, src_dir, dst_dir, mode='RELATIVE', library=texture.image.library)
+        # Do not use if target path relative
+        # filename = repr(filepath)[1:-1]
+        _, filename = split(filename)
+        filename = self.change_ext(filename, texture.image)
```
Comment (paroj) [inline io_ogre/ogre/material.py:224]: yes, we should keep this. It is used by `FORCE_IMAGE_FORMAT` which got lost in the 2.8 conversion

File: io_ogre/util.py [hunk 2]
```diff
@@ -304,6 +303,13 @@ def objects_merge_materials(objs):
             materials.add(mat)
     return materials
 
+def should_export(obj):
```
Comment (paroj) [inline io_ogre/util.py:306]: > Should we use new visibility modes in 2.80:

no idea how they are used. Feel free to add as it suits your usecase

File: io_ogre/ogre/material.py [hunk 3]
```diff
@@ -64,23 +67,37 @@ def dot_material(mat, path, **kwargs):
...
 
 class OgreMaterialGenerator(object):
-    def __init__(self, material, prefix=''):
+    # Texture wrapper attribute names
+    TEXTURE_KEYS = [
+        "base_color_texture",
+        "specular_texture",
+        "roughness_texture",
+        "alpha_texture",
+        "normalmap_texture",    # useless ?
```
Comment (paroj) [inline io_ogre/ogre/material.py:89]: no, quite useful actually. You can tell ogre about it with:
https://ogrecave.github.io/ogre/api/latest/rtss.html#lighting_stage

however, this is out of scope for this PR

File: io_ogre/ogre/material.py [hunk 4]
```diff
@@ -175,11 +202,57 @@ def generate_pass( self, mat, pass_name="" ):
...
+                if texture.projection == 'SPHERE':
+                    self.w.iline('env_map spherical')
+                elif texture.projection == 'FLAT':
+                    self.w.iline('env_map planar')
+                else: 
+                    logging.debug('WARNING: <%s> has a non-UV mapping type (%s) and not picked a proper projection type of: Sphere or Flat' %(texture.name, texture.projection))
+            
+            if texture.translation.x or texture.translation.y:
+                self.w.iword('scroll').real(texture.translation.x).real(texture.translation.y).nl()
+            if texture.rotation.z:
```
Comment (paroj) [inline io_ogre/ogre/material.py:247]: it seems blender never actually supported texture rotation:
https://blender.stackexchange.com/questions/5608/rotate-object-texture

and somebody just picked a free variable for whatever reasons

File: io_ogre/ogre/material.py [hunk 5]
```diff
@@ -175,11 +202,57 @@ def generate_pass( self, mat, pass_name="" ):
...
+                self.w.iword('tex_address_mode').word(TEXTURE_ADDRESS_MODE[exmode]).nl()
 
+            if exmode == 'CLIP':
+                self.w.iword('tex_border_colour').word(texture.color.r).word(texture.color.g).word(texture.color.b).nl()
+            if texture.scale != Vector((1.0, 1.0, 1.0)):
+                self.w.iword('scale').real(1.0 / texture.scale.x).real(1.0 / texture.scale.y).nl()
+            
+            if texture.texcoords == 'Reflection':
+                if texture.projection == 'SPHERE':
+                    self.w.iline('env_map spherical')
```
Comment (paroj) [inline io_ogre/ogre/material.py:239]: this implies "reflection" on the ogre side

**Comment by paroj** [comment]
> use texture name or label instead ?

texture names must be unique in Ogre, so I would pick something that must be unique in blender as well

**Comment by paroj** [comment]
great contribution! thanks. I would merge it as is, because it is a strict improvement of the current state. Unless you want to fix some more things first. So tell me when you are ready.

File: io_ogre/ogre/material.py (hunk 4 above)
Comment (Grodou) [reply io_ogre/ogre/material.py:247]: These values (location, rotation, scale) are retrieved from a Vector Mapping node.
That's how I tested it, and I guess the only way to change an image texture mapping on Blender 2.8x.
What is the unit for ogre rotation ?

**Comment by Grodou** [comment]
Hi,

I use the PrincipledBSDFWrapper for this, as the Wavefront exporter does, but it only works for... the new PrincipledBSDFShader !
Textures are ignored for even a simple Diffuse Shader (also tested with the Wavefront exporter).
I'm trying to find out another way of doing it (there's a ShaderImageTextureWrapper, maybe it can do the job).

So OK, you can merge it, I'll try to fix and clean it later.
And you're welcome, thanks for the review :)

**Comment by funnyflowerpot** [comment]
Thank you both for your hard work on Blender 2.80 support!
